### PR TITLE
First draft of a repository management handbook section.

### PR DESCRIPTION
**Pull request by schlessera**
See #112

### CODE REVIEW

**Comment by schlessera** [comment]
This is still a work in progress. I uploaded it as a pull request now to serve as a base for discussion.

File: repository-management.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
+# Repository Management
+
+## Package Names
+
+Package names for the official `wp-cli` GitHub repository need to be discussed upfront with the [maintainers](https://github.com/orgs/wp-cli/teams/maintainers).
```
Comment (danielbachhuber) [inline repository-management.md]: `s/GitHub repository/GitHub organization`

Can you explain why package naming is important? What's an overview to our decision-making criteria before naming?

Maybe something like this:

> Before creating a new repository, it's important to discuss the repository first. This discussion helps establish consensus on what purpose the codebase serves, and where it lives best in the WP-CLI ecosystem.

File: repository-management.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+Package names for WP-CLI commands must end with the `-command` suffix.
```
Comment (danielbachhuber) [inline repository-management.md]: ... for clarity, and also to distinguish them from the non-command repositories.

File: repository-management.md (hunk 1 above)
Comment (danielbachhuber) [inline repository-management.md]: Also, for commands, it's important to decide whether a new command deserves its own repository or should be included with an existing repository.

File: repository-management.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+_Example: `wp-cli/scaffold-package-command`._
+
+## Package Versions
+
+Package version follow the [Semantic Versioning Specification (SemVer)](http://semver.org/).
```
Comment (danielbachhuber) [inline repository-management.md]: In layman's terms for us, this typically means:

* Patch releases (e.g. `1.0.x`) are used for bug fixes, documentation changes, small enhancements, etc.
* Minor releases  (e.g. `1.x.0`) are used when a new command is introduced to a package, or a major enhancement is made to an existing command.

File: repository-management.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+## Milestones
+
+Open milestones usually point to the next version to be released.
+
+As the release notes are built using the milestones, pull requests that are to be merged need to be set to the upcoming milestone, so that these pull requests will be part of the release notes.
```
Comment (danielbachhuber) [inline repository-management.md]: Both pull requests and issues addressed in a milestone should be assigned to the milestone.

File: repository-management.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+_Example: `command:cli-update`_
+
+#### Scope
+
+The labels that define what scope the current issue/pull request applies to are prefixed with `scope:`.
+
+Used scopes:
+
+* `bootstrap` - Part of the bootstrap process, which loads both WP-CLI as well as WordPress Core.
```
Comment (danielbachhuber) [inline repository-management.md]: For clarity, I'd use `scope:bootstrap` instead of `bootstrap` when referring to specific labels. Doing so the reader doesn't have to connect the docs in their head.

File: repository-management.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+When a pull request was submitted by an external contributor, the [committers](https://github.com/orgs/wp-cli/teams/committers) should be responsive and provide quick feedback to encourage further contributions.
+
+Apart from needing to be approved, pull requests also need to have their tests in a passing state before they can be merged. When both of these conditions are true, any [committer](https://github.com/orgs/wp-cli/teams/committers) can merge the pull requestby:
+1. Ensuring that all applicable labels have been set.
+2. Ensuring that the correct milestone has been set.
+3. Ensuring that the branch is deleted after the merge (if applicable).
+
+## Releases
+
+Releases are scheduled and communicated well in advance by the [maintainers](https://github.com/orgs/wp-cli/teams/maintainers).
```
Comment (danielbachhuber) [inline repository-management.md]: We should distinguish between WP-CLI releases (which are scheduled in advance) and package releases (which happen on an as-needed basis. More specifically, I tag a release for a package when:

* It has a reasonable number of changes that have been sitting for a while (couple weeks or more).
* It has a substantial enhancement (e.g. new command) that's worth getting into the nightly build.

I'll hold off on tagging a release (for a few to several days) if there's an open pull request that makes sense to include in the release.

File: repository-management.md [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+Non-trivial pull requests should be preceded by a related issue that defines the problem to solve and allows for discussion of the most appropriate solution before actually writing code.
+  
+When a pull request was submitted by one of the [committers](https://github.com/orgs/wp-cli/teams/committers), the submitter should set the "Reviewers" for that pull request to `wp-cli/committers` if a general code review is needed, or to one or more specific committer profiles if the expertise of a specific person is needed/wanted.
+
+When a pull request was submitted by an external contributor, the [committers](https://github.com/orgs/wp-cli/teams/committers) should be responsive and provide quick feedback to encourage further contributions.
+
+Apart from needing to be approved, pull requests also need to have their tests in a passing state before they can be merged. When both of these conditions are true, any [committer](https://github.com/orgs/wp-cli/teams/committers) can merge the pull requestby:
+1. Ensuring that all applicable labels have been set.
+2. Ensuring that the correct milestone has been set.
+3. Ensuring that the branch is deleted after the merge (if applicable).
```
Comment (danielbachhuber) [inline repository-management.md]: On the topic of pull requests, another thing I do at the end is re-title the pull request as needed:

image

The ideal pull request title is one that I can more or less copy and paste into release notes.

**Comment by danielbachhuber** [comment]
@schlessera Before you go about revising this pull request (which will likely cause current comments to be hidden), it'd be good to have @wp-cli/committers weigh in with questions and comments.

**Comment by schlessera** [comment]
@wp-cli/committers , any additional feedback before I make changes?

**Comment by gitlost** [comment]
None that I can think of - looks good.

File: repository-management.md [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+Every submitted pull request needs to go through a code review, and needs to be approved by at least one of the [committers](https://github.com/orgs/wp-cli/teams/committers).
+
+Non-trivial pull requests should be preceded by a related issue that defines the problem to solve and allows for discussion of the most appropriate solution before actually writing code.
+  
+When a pull request was submitted by one of the [committers](https://github.com/orgs/wp-cli/teams/committers), the submitter should set the "Reviewers" for that pull request to `wp-cli/committers` if a general code review is needed, or to one or more specific committer profiles if the expertise of a specific person is needed/wanted.
+
+When a pull request was submitted by an external contributor, the [committers](https://github.com/orgs/wp-cli/teams/committers) should be responsive and provide quick feedback to encourage further contributions.
+
+Apart from needing to be approved, pull requests also need to have their tests in a passing state before they can be merged. When both of these conditions are true, any [committer](https://github.com/orgs/wp-cli/teams/committers) can merge the pull requestby:
```
Comment (miya0001) [inline repository-management.md]: /pull requestsby/pull requests by/

File: repository-management.md (hunk 3 above)
Comment (miya0001) [inline repository-management.md]: /Package version follow/Package version follows/

**Comment by danielbachhuber** [comment]
Another note: branches should be deleted after they're merged.

File: repository-management.md [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+## Pull Requests
+
+All code changes go through a pull request.
+
+Every submitted pull request needs to go through a code review, and needs to be approved by at least one of the [committers](https://github.com/orgs/wp-cli/teams/committers).
+
+Non-trivial pull requests should be preceded by a related issue that defines the problem to solve and allows for discussion of the most appropriate solution before actually writing code.
+  
+When a pull request was submitted by one of the [committers](https://github.com/orgs/wp-cli/teams/committers), the submitter should set the "Reviewers" for that pull request to `wp-cli/committers` if a general code review is needed, or to one or more specific committer profiles if the expertise of a specific person is needed/wanted.
```
Comment (Sidsector9) [inline repository-management.md]: @schlessera Shouldn't it be "When a pull request **is** submitted"

File: repository-management.md [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+When a pull request was submitted by an external contributor, the [committers](https://github.com/orgs/wp-cli/teams/committers) should be responsive and provide quick feedback to encourage further contributions.
```
Comment (Sidsector9) [inline repository-management.md]: **is** instead of **was** here too

File: repository-management.md (hunk 9 above)
Comment (schlessera) [reply repository-management.md]: The sentence refers to the one who actually did the submission. The submission itself is in the past then.

But I think `If` would be preferable to `When` in this case.

File: repository-management.md (hunk 1 above)
Comment (schlessera) [reply repository-management.md]: https://github.com/wp-cli/handbook/pull/127/commits/1deaf59f20647653ae2155165db88ebd46bb9cc6 & https://github.com/wp-cli/handbook/pull/127/commits/7aebd670232e4516648e7a0ebdbdc265920f0e0f

File: repository-management.md (hunk 2 above)
Comment (schlessera) [reply repository-management.md]: https://github.com/wp-cli/handbook/pull/127/commits/51d8b06cf4c4953a2a1e5d2b4bf609a67818ddc0

File: repository-management.md (hunk 3 above)
Comment (schlessera) [reply repository-management.md]: https://github.com/wp-cli/handbook/pull/127/commits/20a23d4c18f260317ed24ed70245463b60537ff8 & https://github.com/wp-cli/handbook/pull/127/commits/e9efb1a4ce74687863cb29debb0b12f6a23d6696

File: repository-management.md (hunk 4 above)
Comment (schlessera) [reply repository-management.md]: https://github.com/wp-cli/handbook/pull/127/commits/b2a3d71cae8daf696a6e0269deaec353e35502d8

File: repository-management.md (hunk 5 above)
Comment (schlessera) [reply repository-management.md]: https://github.com/wp-cli/handbook/pull/127/commits/95f0f62007eea934fc4f47269aae81a4107104be

File: repository-management.md (hunk 6 above)
Comment (schlessera) [reply repository-management.md]: https://github.com/wp-cli/handbook/pull/127/commits/bd9fecac25fc5aa3d9d3a0923711f998e2196bc9

File: repository-management.md (hunk 7 above)
Comment (schlessera) [reply repository-management.md]: https://github.com/wp-cli/handbook/pull/127/commits/c286eb6eb0bcdf82bdf2a9cb36eb7cb9ba40218b

File: repository-management.md (hunk 8 above)
Comment (schlessera) [reply repository-management.md]: https://github.com/wp-cli/handbook/pull/127/commits/a2634968cf82f38d2925016498e754aecc4e0c1f

File: repository-management.md (hunk 9 above)
Comment (schlessera) [reply repository-management.md]: https://github.com/wp-cli/handbook/pull/127/commits/a19d28f2a3551852272db5060852b8b43e78f745

File: repository-management.md [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+_Example: `command:cli-update`_
+
+#### Scope
+
+The labels that define what scope the current issue/pull request applies to are prefixed with `scope:`.
+
+Used scopes:
+
+* `scope:bootstrap` - Part of the bootstrap process, which loads both WP-CLI as well as WordPress Core.
+* `scope:command` - Part of an individual command.
```
Comment (danielbachhuber) [inline repository-management.md]: Is `scope:command` redundant when we have `command:<name>`?

File: repository-management.md [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+
+## Milestones
+
+Open milestones usually point to the next version to be released.
+
+As the release notes are built using the milestones, both pull requests and issues that are to be merged need to be set to the upcoming milestone, so that these pull requests will be part of the release notes.
+
+## Labels
+
+Labels are an important organizational tool to communicate the state and progress of the work being done. [Committers](https://github.com/orgs/wp-cli/teams/committers) are required to keep these up-to-date.
```
Comment (danielbachhuber) [inline repository-management.md]: Might be worth saying something along the lines of:

> Keeping issues and pull requests accurately labeled helps us make sense of issue tracker usage in aggregate and more easily search closed issues and pull requests for past discussion.

Comment (danielbachhuber) [reply repository-management.md]: e.g. better explaining _why_

File: repository-management.md [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+
+### WP-CLI Releases
+
+WP-CLI releases are scheduled and communicated well in advance by the [maintainers](https://github.com/orgs/wp-cli/teams/maintainers).
+
+When the release is ready, it is being prepared and executed according to the [release checklist](release-checklist.md) by one of the [maintainers](https://github.com/orgs/wp-cli/teams/maintainers).
+
+### Package Releases
+
+Package releases happen on an as-needed basis. A package is tagged for a release when:
```
Comment (danielbachhuber) [inline repository-management.md]: Let's keep package releases to the domain of the maintainers for now. Doing so will let us do a final review of all code in the release before it gets tagged.

File: repository-management.md (hunk 11 above)
Comment (schlessera) [reply repository-management.md]: I was also wondering the same thing. Do you think I should remove it, provided we make a consistent habit of using the `command:*` labels?

I'll add these as an additional section then.

File: repository-management.md (hunk 12 above)
Comment (schlessera) [reply repository-management.md]: https://github.com/wp-cli/handbook/pull/127/commits/31e75dd539e5f045e97c081247709ed9a651ac77

File: repository-management.md (hunk 11 above)
Comment (danielbachhuber) [reply repository-management.md]: > Do you think I should remove it, provided we make a consistent habit of using the `command:*` labels?

Yes, I think so.

File: repository-management.md (hunk 13 above)
Comment (schlessera) [reply repository-management.md]: https://github.com/wp-cli/handbook/pull/127/commits/01a7f496dc2d79a5a6cef378baf276262b09ce71

File: repository-management.md [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+The labels that defined what state a given issue/pull request is in are prefixed with `state:`.
+
+Used states:
+
+* `state:unconfirmed` - The bug/problem in the issue could not be replicated yet or might be related to the reporter's environment.
+* `state:unsupported` - The issue is outside of the scope of a bug report and cannot be supported on GitHub. The reporter should be pointed towards one of the [support channels](http://wp-cli.org/#support).
+
+#### Command
+
+The labels that define what specific command a given issue/pull request is related to are prefix with `command:`.
```
Comment (gitlost) [inline repository-management.md]: Typo "prefix" -> "prefixed".

File: repository-management.md [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+
+## Pull Requests
+
+All code changes go through a pull request.
+
+Every submitted pull request needs to go through a code review, and needs to be approved by at least one of the [committers](https://github.com/orgs/wp-cli/teams/committers).
+
+Non-trivial pull requests should be preceded by a related issue that defines the problem to solve and allows for discussion of the most appropriate solution before actually writing code.
+  
+If a pull request was submitted by one of the [committers](https://github.com/orgs/wp-cli/teams/committers), the submitter should set the "Reviewers" for that pull request to `wp-cli/committers` if a general code review is needed, or to one or more specific committer profiles if the expertise of a specific person is needed/wanted.
```
Comment (gitlost) [inline repository-management.md]: I'm with @Sidsector9 here, "is submitted" sounds better (to my delicate ears anyway), ["present perfect"](https://en.wikipedia.org/wiki/Uses_of_English_verb_forms#Present_perfect) or somethink...

File: repository-management.md [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+
+If a pull request was submitted by an external contributor, the [committers](https://github.com/orgs/wp-cli/teams/committers) should be responsive and provide quick feedback to encourage further contributions.
```
Comment (gitlost) [inline repository-management.md]: Ditto re "is submitted".

File: repository-management.md [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+
+Apart from needing to be approved, pull requests also need to have their tests in a passing state before they can be merged. When both of these conditions are true, any [committer](https://github.com/orgs/wp-cli/teams/committers) can merge the pull request by:
+1. Ensuring that all applicable labels have been set.
+2. Ensuring that the correct milestone has been set.
+3. Ensuring that the branch is deleted after the merge (if applicable).
+4. Adapting the title of the pull-request as needed. The ideal pull request title is one that we can copy and paste into release notes as is.
```
Comment (gitlost) [inline repository-management.md]: First use of "we", maybe keep it impersonal "can be copied and pasted".

File: repository-management.md [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+
+## Releases
+
+### WP-CLI Releases
+
+WP-CLI releases are scheduled and communicated well in advance by the [maintainers](https://github.com/orgs/wp-cli/teams/maintainers).
+
+When the release is ready, it is being prepared and executed according to the [release checklist](release-checklist.md) by one of the [maintainers](https://github.com/orgs/wp-cli/teams/maintainers).
```
Comment (gitlost) [inline repository-management.md]: The "being" isn't needed here.

Comment (schlessera) [reply repository-management.md]: https://github.com/wp-cli/handbook/pull/127/commits/9939df8dbd2cc086de2e72998c121ee0283480ed

File: repository-management.md (hunk 17 above)
Comment (schlessera) [reply repository-management.md]: https://github.com/wp-cli/handbook/pull/127/commits/263be425d6ff582601d24334a43c5977d5de4dbf

File: repository-management.md (hunk 15 above)
Comment (schlessera) [reply repository-management.md]: https://github.com/wp-cli/handbook/pull/127/commits/334ee9f4856fd2aea5a1b2b5c58823ef9b6a90f7

File: repository-management.md (hunk 14 above)
Comment (schlessera) [reply repository-management.md]: https://github.com/wp-cli/handbook/pull/127/commits/8c7213fc7c590f4b39b435eefbea1c499631a8ae

File: repository-management.md (hunk 9 above)
Comment (Sidsector9) [reply repository-management.md]: Yes, `If` would be preferable. Because the sentence sounds odd when it begins with `When a pull request was submitted... submitter should set the...`
